File: kindergartens/Probabilities.py
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 27 13:19:30 2019

@author: W7SP1
"""

import pandas as pd
import os
import locale
from locale import atof
import numpy
from numpy import random
import timeit



locale.setlocale(locale.LC_CTYPE, 'bulgarian')

os.chdir("C:/Documents/GitHub/Kindergarten-Scraping/kindergartens") 

FullList = pd.read_csv("waiting.csv",encoding='windows-1251') 

start = timeit.default_timer()
for k in range(0, 1):
    FullList = FullList.loc[FullList.born == 2017, ['id', 'kg','points','places','tail','born']]
    FullList['rn'] = numpy.random.uniform(size=len(FullList))
    
    dat = FullList[FullList['tail']== "Социални"].sort_values(['points', 'rn'], ascending=[0, 1])
    dat['admitted'] = 0
    
    places = dat[['kg','places']].drop_duplicates()
    places.reset_index(inplace=True,drop = True)
    dat.reset_index(inplace=True,drop = True)
    admitted = list()
    
    
    for i in range(0, len(dat)):
        if (places.loc[places['kg']==dat.loc[i,'kg'],'places'] >0).item() & (dat.loc[i,'id'] not in admitted):
            dat.loc[i, "admitted"] = 1
            places.loc[places['kg']==dat.loc[i,'kg'],'places'] = places.loc[places['kg']==dat.loc[i,'kg'],'places'] - 1
            admitted.append(dat.loc[i,'id'])
        
    admitted_s = dat.loc[dat['admitted']==1,['id','kg','points']].drop_duplicates()
    
    dat = FullList[FullList['tail']== "Общи"].sort_values(['points', 'rn'], ascending=[0, 1])
    dat['admitted'] = 0
    
    places = pd.concat([dat[['kg','places']].drop_duplicates(),places])
    places=places.groupby(['kg'])['places'].sum().reset_index()
    dat.reset_index(inplace=True,drop = True)
    
    for i in range(0, len(dat)):
        if (places.loc[places['kg']==dat.loc[i,'kg'],'places'] >0).item() & (dat.loc[i,'id'] not in admitted):
            dat.loc[i, "admitted"] = 1
            places.loc[places['kg']==dat.loc[i,'kg'],'places'] = places.loc[places['kg']==dat.loc[i,'kg'],'places'] - 1
            admitted.append(dat.loc[i,'id'])
     
    admitted_o = dat.loc[dat['admitted']==1,['id','kg','points']].drop_duplicates()
    ThisRunAdmitted=pd.concat([admitted_s,admitted_o])
    ThisRunAdmitted['run'] =1
    
        
    try:
        FullAdmitted
    except:
         FullAdmitted = ThisRunAdmitted
    else:
         FullAdmitted = pd.concat([FullAdmitted,ThisRunAdmitted])
    
    del([ThisRunAdmitted,admitted_s, admitted_o, places])
    


end = timeit.default_timer()
end-start

File: kindergartens/Probabilities.R
Sys.setlocale("LC_CTYPE", "bulgarian")
##### SAVE WITH ENCODING MS-CYRL

library(dplyr)
library(snow)
library(Hmisc)
library(parallel)
library(doParallel)

numCores <- detectCores()
numCores

cl <- makeCluster(numCores)
registerDoParallel(cl)


setwd("C:/Documents/GitHub/Kindergarten-Scraping/kindergartens")

FullList <- read.csv("waiting.csv",stringsAsFactors = F)

set.seed(100)

runonce <- function(k) {
  FullList <- FullList %>% filter (born==2017) %>% select(id, kg, points, places, tail,born)  
  FullList$rn <- runif(nrow(FullList))
  
  dat <- FullList %>% filter (tail == "Социални") %>% arrange(desc (points), rn)
  dat$admitted = 0
  
  places <- dat %>% select(kg, places) %>% distinct()
  admitted <- c()
  
  
  for (i in (1:nrow(dat))) {
    if (places$places[places$kg == dat[i, "kg"]] > 0 & !dat[i, "id"] %in% admitted) {
      dat[i, "admitted"] <- 1
      places$places[places$kg == dat[i, "kg"]] <- places$places[places$kg == dat[i, "kg"]] - 1
      admitted <- c(admitted, dat[i, "id"])
    }
  }
  
  
  admitted_s <- dat %>% filter(admitted==1) %>% select(id, kg,points) %>% distinct()
  
  dat <- FullList %>% filter (tail == "Общи") %>% arrange(desc (points), rn)
  dat$admitted = 0
  
  places <- rbind(dat %>% select(kg, places) %>% distinct(),places)
  places <- places %>% group_by (kg) %>% summarise(places = sum(places)) %>% ungroup
  
  for (i in (1:nrow(dat))) {
    if (places$places[places$kg == dat[i, "kg"]] > 0 & !dat[i, "id"] %in% admitted) {
      dat[i, "admitted"] <- 1
      places$places[places$kg == dat[i, "kg"]] <- places$places[places$kg == dat[i, "kg"]] - 1
      admitted <- c(admitted, dat[i, "id"])
    }
  }

  
  admitted_o <- dat %>% filter(admitted==1) %>% select(id, kg,points) %>% distinct()
  
  ThisRunAdmitted <- data.frame("run" = k, rbind(admitted_s,admitted_o))

  return(ThisRunAdmitted)

  
  rm(ThisRunAdmitted, admitted_s, admitted_o, places)
}


#system.time(x<-apply(1:4,runonce))
clusterExport(cl,list("FullList"))

system.time(FullAdmitted <-foreach (k = 1:100, .combine = rbind) %dopar% {
  Sys.setlocale("LC_CTYPE", "bulgarian")
  library(dplyr)
  runonce(k)}
  )


stopCluster(cl) 


runs <- max(FullAdmitted$run)
chances <- FullAdmitted %>% group_by(id) %>% summarise(proba = length(id)/runs)
#x<-FullAdmitted %>% group_by(kg) %>% summarise(min(points))

chances %>% filter(id ==17004985)
FullAdmitted %>% filter(id ==17004985) %>% group_by(kg) %>% summarise(num = length(kg)/runs)
FullList %>% filter(id ==17002770) %>% select( kg,tail,points)
FullList %>%filter(!id %in% chances$id & born==2017) %>% summarise(num = n_distinct(id))
chances %>% summarise(num = n_distinct(id))
dat %>% filter(id ==17002770)
